### PR TITLE
Support vector similarity

### PR DESCRIPTION
**Pull request by Avital-Fine**

### CODE REVIEW

**Review by sazzad16 — CHANGES_REQUESTED** [review]
```
[ERROR] Tests run: 33, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 0.085 s <<< FAILURE! - in redis.clients.jedis.modules.search.SearchTest
[ERROR] testQueryParams(redis.clients.jedis.modules.search.SearchTest)  Time elapsed: 0.006 s  <<< FAILURE!
java.lang.AssertionError: expected:<2> but was:<0>
	at redis.clients.jedis.modules.search.SearchTest.testQueryParams(SearchTest.java:456)
```

File: src/main/java/redis/clients/jedis/search/Schema.java [hunk 1]
```diff
@@ -273,4 +278,34 @@ public String toString() {
           + ", separator='" + separator + "'}";
     }
   }
+
+  public static class VectorField extends Field {
+
+    public enum VectorAlgo {
```
Comment (Avital-Fine) [inline src/main/java/redis/clients/jedis/search/Schema.java:284]: @sazzad16 
I'm not sure.. maybe String is better in this case?

File: src/test/java/redis/clients/jedis/modules/search/SearchTest.java [hunk 2]
```diff
@@ -443,6 +438,42 @@ public void testQueryFlags() throws Exception {
     assertEquals(0, res.getTotalResults());
   }
 
+  @Test
+  public void testVectorSimilarity() {
```
Comment (Avital-Fine) [inline src/test/java/redis/clients/jedis/modules/search/SearchTest.java:442]: taken from one of the module's tests : https://github.com/RediSearch/RediSearch/blob/e139248127a17c81cddad840fca2fd99e80f5f41/tests/pytests/test_search_params.py#L312

File: src/main/java/redis/clients/jedis/search/Query.java [hunk 3]
```diff
@@ -291,6 +294,15 @@ public void addParams(CommandArguments args) {
 //      args.set(returnCountIndex, Protocol.toByteArray(returnCount));
       returnCountObject.setRaw(Protocol.toByteArray(returnCount));
     }
+
+    if (_params != null && _params.size() > 0) {
+      args.add(SearchKeyword.PARAMS.getRaw());
+      args.add(_params.size() * 2);
+      for (Map.Entry<String, Object> entry : _params.entrySet()){
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/search/Query.java:301]:
```suggestion
      for (Map.Entry<String, Object> entry : _params.entrySet()) {
```

File: src/main/java/redis/clients/jedis/search/Schema.java [hunk 4]
```diff
@@ -273,4 +278,33 @@ public String toString() {
...
+      FLAT,
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/search/Schema.java:285]: Add a test for FLAT.

File: src/main/java/redis/clients/jedis/search/Schema.java [hunk 5]
```diff
@@ -99,6 +100,11 @@ public Schema addSortableTagField(String name, String separator) {
     return this;
   }
 
+  public Schema addVectorField(String name, VectorField.VectorAlgo algorithm, List<String> attributes) {
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/search/Schema.java:103]: Add 2 more methods: addFlatVectorField, addHNSWVectorField.

**Review by sazzad16 — CHANGES_REQUESTED** [review]

File: src/test/java/redis/clients/jedis/modules/search/SearchTest.java [hunk 6]
```diff
@@ -443,6 +438,42 @@ public void testQueryFlags() throws Exception {
...
+    attr.add("L2");
+    Schema sc = new Schema().addVectorField("v", Schema.VectorField.VectorAlgo.HNSW, attr);
+    assertEquals("OK", client.ftCreate(index, IndexOptions.defaultOptions(), sc));
+
+    client.hset("a", "v", "aaaaaaaa");
+    client.hset("b", "v", "aaaabaaa");
+    client.hset("c", "v", "aaaaabaa");
+
+    Query query =  new Query("*=>[KNN 2 @v $vec]").addParam("vec", "aaaaaaaa")
+            .setSortBy("__v_score", true).returnFields("__v_score");
```
Comment (sazzad16) [inline src/test/java/redis/clients/jedis/modules/search/SearchTest.java:458]:
```suggestion
    Query query =  new Query("*=>[KNN 2 @v $vec]")
        .addParam("vec", "aaaaaaaa")
        .setSortBy("__v_score", true)
        .returnFields("__v_score");
```

**Review by sazzad16 — CHANGES_REQUESTED** [review]

File: src/main/java/redis/clients/jedis/search/Schema.java (hunk 5 above)
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/search/Schema.java:103]: `Map<String, Object> attributes` (instead of `List<String> attributes`)

**Review by sazzad16 — CHANGES_REQUESTED** [review]

File: src/main/java/redis/clients/jedis/search/Schema.java (hunk 4 above)
Comment (Avital-Fine) [reply src/main/java/redis/clients/jedis/search/Schema.java:285]: The same test is working with FLAT. Do you want me to do duplicate it? :sweat_smile:

Comment (sazzad16) [reply src/main/java/redis/clients/jedis/search/Schema.java:297]: Yes

File: src/test/java/redis/clients/jedis/modules/search/SchemaTest.java [hunk 7]
```diff
@@ -22,10 +25,12 @@ public void printSchemaTest() throws Exception {
         .addSortableTagField(GENRE, ",")
         .addSortableNumericField(RELEASE_YEAR)
         .addSortableNumericField(RATING)
-        .addSortableNumericField(VOTES);
+        .addSortableNumericField(VOTES)
+        .addVectorField(VECTOR, Schema.VectorField.VectorAlgo.HNSW, new HashMap<>());
```
Comment (sazzad16) [inline src/test/java/redis/clients/jedis/modules/search/SchemaTest.java:29]: Collections.emptyMap()

**Review by sazzad16 — APPROVED** [review]

**Pull request merged**